frontend/class-public.php

```diff
@@ -388,16 +388,53 @@ public function process_volunteer_sign_up(){
 	 */
 	public function send_email_reminder( $opp_id ){
 
+		$opp = new WI_Volunteer_Management_Opportunity( $opp_id );
+
+		$reminder_email_already_sent = $this->has_reminder_email_already_sent( $opp );
+		if( $reminder_email_already_sent === true ){
+			return false;
+		}
+
 		$data_array = array(
 			'post_id' => $opp_id,
 			'user_id' => 0,
 		);
 
-		$opp 	= new WI_Volunteer_Management_Opportunity( $opp_id );
 		$email 	= new WI_Volunteer_Management_Email( $opp );
 		$email->send_volunteer_reminder_email();
 		$email->store_volunteer_email( $data_array );
+	}
+
+	/**
+	 * Check if a reminder email has already been sent by the system within the last 5 hours.
+	 *
+	 * While send_email_reminder() should only run one time through cron, in certain caching
+	 * situations the same cron event was run multiple times, which triggered the automated
+	 * reminder email to send multiple times. Since the get_rsvp_emails() method uses
+	 * $wpdb->get_results() which is not cached, this method should prevent the reminder
+	 * email from being sent multiple times.
+	 * 
+	 * @param  object  $opp The current volunteer opportunity to check.
+	 * @return boolean      Whether an automated reminder email has been sent within the last 5 hours.
+	 */
+	public function has_reminder_email_already_sent( $opp ){
+
+		$current_time 				= current_time( 'timestamp' );
+		$five_hours_in_seconds 		= HOUR_IN_SECONDS * 5; // 18,000 seconds
+		$five_hours_ago_in_seconds 	= $current_time - $five_hours_in_seconds;
+		$emails_sent				= $opp->get_rsvp_emails();
+
+		foreach( $emails_sent as $email ){
+
+			$email_sent_time_in_seconds = strtotime( $email->time );
+
```

Comment:
What is the significance of checking against a time? Couldn't we just check to see that $email_sent_time_in_seconds (or $email->time) is set/not blank which would be indicator enough that the email was already sent?

Reply:
Oh, this was an interesting question that I wrestled with a lot. So, the options are this:

1. Check if a reminder email has been sent at all.
1. Check if a reminder email has been sent in the last five hours.

I decided to check the last five hours in case of a situation where a reminder email has already been sent, then the volunteer opportunity is actually moved to a new date and time, let's say a few months in the future. If we checked for a reminder email at any time then they would never get one a few months down the road. Does that make sense? I'd love to hear your thoughts since I went back and forth.

Reply:
Oh yes, for sure. That does make a ton of sense.

That actually did cross my mind as well - for instance if instead of leaving the Opportunity in the past once it's over and creating a new Opportunity, the admin went in and changed the date and time to a future date and time. Then we would still want the Reminder Emails to send so we wouldn't want the check to be only if the email has already sent.

I'm on board with this approach! Thanks for the explanation.

Reply:
Awesome, thanks a ton Dan.